File: eco_restoration_shard/cyboquatics/2026-08-17-b-residual-corridor/cpp/canal_kalman_filter.cpp
#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

struct Step {
    double measurement;
    bool available;
};

struct FilterState {
    double estimate;
    double covariance;
};

struct Output {
    double innovation;
    double innovation_variance;
    double normalized_innovation;
    double gain;
    bool updated;
};

static bool finite(double value) {
    return std::isfinite(value);
}

static Output update(
    FilterState& state,
    double process_variance,
    double measurement_variance,
    double h,
    const Step& step
) {
    if (process_variance < 0.0 || measurement_variance <= 0.0 || h == 0.0) {
        throw std::invalid_argument("require Q >= 0, R > 0, and non-zero H");
    }

    const double prior_estimate = state.estimate;
    const double prior_covariance = state.covariance + process_variance;

    if (!step.available) {
        state.estimate = prior_estimate;
        state.covariance = prior_covariance;
        return {NAN, NAN, NAN, 0.0, false};
    }

    const double innovation = step.measurement - h * prior_estimate;
    const double innovation_variance = h * prior_covariance * h + measurement_variance;
    const double gain = prior_covariance * h / innovation_variance;

    state.estimate = prior_estimate + gain * innovation;
    state.covariance = std::max(0.0, (1.0 - gain * h) * prior_covariance);

    return {
        innovation,
        innovation_variance,
        innovation / std::sqrt(innovation_variance),
        gain,
        true
    };
}

static double lag1_autocorrelation(const std::vector<double>& values) {
    if (values.size() < 3) {
        return NAN;
    }

    double mean = 0.0;
    for (double value : values) {
        mean += value;
    }
    mean /= static_cast<double>(values.size());

    double numerator = 0.0;
    double denominator = 0.0;
    for (std::size_t index = 0; index < values.size(); ++index) {
        const double centered = values[index] - mean;
        denominator += centered * centered;
        if (index > 0) {
            numerator += centered * (values[index - 1] - mean);
        }
    }

    return denominator <= 0.0 ? NAN : numerator / denominator;
}

int main(int argc, char** argv) {
    if (argc < 8 || ((argc - 4) % 2 != 0)) {
        std::cerr
            << "usage: " << argv[0]
            << " <x0> <P0> <Q> <R> <H> <measurement_or_na> <available_0_or_1> "
            << "[<measurement_or_na> <available_0_or_1> ...]\n";
        return 64;
    }

    try {
        FilterState state{std::stod(argv[1]), std::stod(argv[2])};
        const double process_variance = std::stod(argv[3]);
        const double measurement_variance = std::stod(argv[4]);
        const double h = std::stod(argv[5]);

        if (state.covariance < 0.0) {
            throw std::invalid_argument("P0 must be non-negative");
        }

        std::vector<double> normalized_innovations;
        for (int index = 6, step_index = 0; index < argc; index += 2, ++step_index) {
            const bool available = std::stoi(argv[index + 1]) == 1;
            const Step step{available ? std::stod(argv[index]) : 0.0, available};
            const Output result = update(state, process_variance, measurement_variance, h, step);

            std::cout << std::fixed << std::setprecision(8)
                      << "step=" << step_index
                      << " updated=" << (result.updated ? 1 : 0)
                      << " estimate=" << state.estimate
                      << " covariance=" << state.covariance
                      << " gain=" << result.gain;

            if (result.updated) {
                normalized_innovations.push_back(result.normalized_innovation);
                std::cout << " residual=" << result.innovation
                          << " S=" << result.innovation_variance
                          << " normalized_residual=" << result.normalized_innovation;
            }
            std::cout << '\n';
        }

        const double lag1 = lag1_autocorrelation(normalized_innovations);
        std::cout << "available_innovation_count=" << normalized_innovations.size() << '\n';
        std::cout << "lag1_normalized_residual_autocorrelation=";
        if (finite(lag1)) {
            std::cout << std::fixed << std::setprecision(8) << lag1 << '\n';
        } else {
            std::cout << "UNAVAILABLE\n";
        }
        std::cout << "whiteness_screen="
                  << (finite(lag1) && std::abs(lag1) <= 0.20 ? "PASS_SCREEN_ONLY" : "REVIEW_MODEL_OR_COLLECT_DATA")
                  << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
        return 65;
    }
}

File: eco_restoration_shard/cyboquatics/2026-08-17-b-residual-corridor/java/CanalKalmanFilter.java
import java.util.ArrayList;
import java.util.List;

public final class CanalKalmanFilter {
    private CanalKalmanFilter() {}

    private record State(double estimate, double covariance) {}
    private record Result(State state, double residual, double innovationVariance, double normalizedResidual, double gain, boolean updated) {}

    private static Result update(
        State state,
        double processVariance,
        double measurementVariance,
        double h,
        double measurement,
        boolean available
    ) {
        if (processVariance < 0.0 || measurementVariance <= 0.0 || h == 0.0) {
            throw new IllegalArgumentException("require Q >= 0, R > 0, and non-zero H");
        }

        double priorEstimate = state.estimate();
        double priorCovariance = state.covariance() + processVariance;

        if (!available) {
            return new Result(new State(priorEstimate, priorCovariance), Double.NaN, Double.NaN, Double.NaN, 0.0, false);
        }

        double residual = measurement - h * priorEstimate;
        double innovationVariance = h * priorCovariance * h + measurementVariance;
        double gain = priorCovariance * h / innovationVariance;
        double posteriorEstimate = priorEstimate + gain * residual;
        double posteriorCovariance = Math.max(0.0, (1.0 - gain * h) * priorCovariance);
        return new Result(
            new State(posteriorEstimate, posteriorCovariance),
            residual,
            innovationVariance,
            residual / Math.sqrt(innovationVariance),
            gain,
            true
        );
    }

    private static double lagOneAutocorrelation(List<Double> values) {
        if (values.size() < 3) {
            return Double.NaN;
        }

        double mean = values.stream().mapToDouble(Double::doubleValue).average().orElse(Double.NaN);
        double numerator = 0.0;
        double denominator = 0.0;
        for (int index = 0; index < values.size(); index++) {
            double centered = values.get(index) - mean;
            denominator += centered * centered;
            if (index > 0) {
                numerator += centered * (values.get(index - 1) - mean);
            }
        }
        return denominator <= 0.0 ? Double.NaN : numerator / denominator;
    }

    public static void main(String[] args) {
        if (args.length < 8 || (args.length - 6) % 2 != 0) {
            System.err.println(
                "usage: CanalKalmanFilter <x0> <P0> <Q> <R> <H> <measurement_or_zero> <available_0_or_1> "
                + "[<measurement_or_zero> <available_0_or_1> ...]"
            );
            System.exit(64);
        }

        try {
            State state = new State(Double.parseDouble(args[0]), Double.parseDouble(args[1]));
            double processVariance = Double.parseDouble(args[2]);
            double measurementVariance = Double.parseDouble(args[3]);
            double h = Double.parseDouble(args[4]);

            if (state.covariance() < 0.0) {
                throw new IllegalArgumentException("P0 must be non-negative");
            }

            List<Double> innovations = new ArrayList<>();
            for (int index = 5, step = 0; index < args.length; index += 2, step++) {
                boolean available = Integer.parseInt(args[index + 1]) == 1;
                double measurement = available ? Double.parseDouble(args[index]) : 0.0;
                Result result = update(state, processVariance, measurementVariance, h, measurement, available);
                state = result.state();

                System.out.printf(
                    "step=%d updated=%d estimate=%.8f covariance=%.8f gain=%.8f%n",
                    step, result.updated() ? 1 : 0, state.estimate(), state.covariance(), result.gain()
                );

                if (result.updated()) {
                    innovations.add(result.normalizedResidual());
                    System.out.printf(
                        "residual=%.8f S=%.8f normalized_residual=%.8f%n",
                        result.residual(), result.innovationVariance(), result.normalizedResidual()
                    );
                }
            }

            double lagOne = lagOneAutocorrelation(innovations);
            System.out.println("available_innovation_count=" + innovations.size());
            System.out.println(
                "lag1_normalized_residual_autocorrelation=" +
                (Double.isFinite(lagOne) ? String.format("%.8f", lagOne) : "UNAVAILABLE")
            );
            System.out.println(
                "whiteness_screen=" +
                (Double.isFinite(lagOne) && Math.abs(lagOne) <= 0.20
                    ? "PASS_SCREEN_ONLY" : "REVIEW_MODEL_OR_COLLECT_DATA")
            );
        } catch (IllegalArgumentException exception) {
            System.err.println("error: " + exception.getMessage());
            System.exit(65);
        }
    }
}
